frame/staking-rewards/src/lib.rs

```diff
@@ -1051,58 +1118,128 @@ fn update_rewards_pool<T: Config>(
+
+		// saturating is safe here since these values are checked against max_rewards anyways, which
+		// is <= u128::MAX
+		let newly_accumulated_rewards =
+			u128::saturating_mul(releasable_periods_surpassed, reward.reward_rate.amount.into());
+		let new_total_rewards =
+			newly_accumulated_rewards.saturating_add(reward.total_rewards.into());
+
+		// u64::MAX is roughly 584.9 billion years in the future, so saturating at that should be ok
+		let last_updated_timestamp = reward.last_updated_timestamp.saturating_add(
+			reward_rate_period_seconds
+				.get()
+				.saturating_mul(releasable_periods_surpassed.saturated_into::<u64>()),
+		);
+
+		// TODO(benluelo): This can probably be simplified, review the period calculations
+		if u128::saturating_add(new_total_rewards, reward.total_dilution_adjustment.into()) <=
+			reward.max_rewards.into()
+		{
+			let balance_on_hold = T::Assets::balance_on_hold(reward.asset_id, pool_account);
+			if !balance_on_hold.is_zero() && balance_on_hold >= newly_accumulated_rewards.into() {
+				T::Assets::release(
+					reward.asset_id,
+					pool_account,
+					newly_accumulated_rewards.into(),
+					false, // not best effort, entire amount must be released
+				)
+				.expect("funds should be avaliable to release based on previous check; qed;");
+
+				reward.total_rewards = new_total_rewards.into();
+				reward.last_updated_timestamp = last_updated_timestamp;
+				Ok(())
 			} else {
-				let new_total_rewards = u128::from(periods_surpassed)
-					.saturating_mul(reward.reward_rate.amount.into())
-					.saturating_add(reward.total_rewards.into());
-
-				let last_updated_timestamp = reward
-					.last_updated_timestamp
-					.add(periods_surpassed.mul(reward_rate_period_seconds.get()));
-
-				if new_total_rewards <= reward.max_rewards.into() {
-					Ok(Reward {
-						total_rewards: new_total_rewards.into(),
-						last_updated_timestamp,
-						..reward
-					})
-				} else if reward.total_rewards < reward.max_rewards {
-					Err(RewardAccumulationCalculationError::MaxRewardsAccumulated(Reward {
-						total_rewards: reward.max_rewards,
-						last_updated_timestamp,
-						..reward
-					}))
-				} else {
-					Err(RewardAccumulationCalculationError::MaxRewardsAccumulatedPreviously(reward))
-				}
+				Err(RewardAccumulationCalculationError::RewardsPotEmpty)
 			}
-		},
-		Err(_) => Err(RewardAccumulationCalculationError::BackToTheFuture(reward)),
+		} else if reward.total_rewards.saturating_add(reward.total_dilution_adjustment) <
+			reward.max_rewards
+		{
+			// if the new total rewards are less than or equal to the max rewards AND the current
+			// total rewards are less than the max rewards (i.e. the newly accumulated rewards is
+			// less than the the amount that would be accumulated based on the periods surpassed),
+			// then release *up to* the max rewards
+			//
+			// REVIEW(benluelo): Should max_rewards be max_periods instead? Currently, the
```

Comment:
Given how this is evolving, my idea is that we can completely remove the `max_rewards` field. Having already a reward pot means that there is already a way to control the amounts being rewarded. We can make it in the future so that if the reward pot account is empty the pool stops being active.

Reply:
I think that would make more sense - would we want to then restrict who can add to the pool?

Reply:
(I will create a ticket to review this)

Reply:
It should be ok to have anyone add to the pool and keep the pool alive actually, that's basically a gauge/bribe protocol :)